### PR TITLE
Logs WAL marker encoding and atomic swap

### PR DESCRIPTION
**Pull request by thepalbi**
#### PR Description

This PR into https://github.com/grafana/agent/pull/5590 proposes a marker file format for `loki.write`. Also, it introduces a lib that allows us to write marker's atomically both in UNIX and Windows OSes.

#### PR Checklist

- [ ] CHANGELOG.md updated
- [ ] Documentation added
- [ ] Tests updated
- [ ] Config converters updated

### CODE REVIEW

File: component/common/loki/client/internal/marker_encoding.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	// first check that read byte stream has expected length
+	if len(bs) != 14 {
+		return 0, fmt.Errorf("bad length %d", len(bs))
+	}
+
+	// check CRC first
+	expectedCrc := crc32.ChecksumIEEE(bs[0:10])
+	gotCrc := binary.BigEndian.Uint32(bs[len(bs)-4:])
+	if expectedCrc != gotCrc {
+		return 0, fmt.Errorf("bad checksum")
```
Comment (ptodev) [inline component/common/loki/client/internal/marker_encoding.go:44]:
```suggestion
		return 0, fmt.Errorf("corrupted WAL marker")
```

File: component/common/loki/client/internal/marker_encoding.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	// check expected version header
+	header := bs[:2]
+	if !(header[0] == markerHeaderV1[0] && header[1] == markerHeaderV1[1]) {
+		return 0, fmt.Errorf("wrong marker header")
```
Comment (ptodev) [inline component/common/loki/client/internal/marker_encoding.go:50]:
```suggestion
		return 0, fmt.Errorf("wrong WAL marker header")
```

**Review by ptodev — APPROVED** [review]
Looks good, thank you! I just added a few minor suggestions.

BTW Prometheus also uses CRC32:
https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/README.md

**Comment by thepalbi** [comment]
> Looks good, thank you! I just added a few minor suggestions.
> 
> BTW Prometheus also uses CRC32: https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/README.md

Awesome, didn't know they were using CRC too. Good choice then!

**Pull request merged**